src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs

```diff
@@ -56,7 +56,20 @@ public ControllerActionInvokerState GetState(ControllerContext controllerContext
             Entry cacheEntry;
             if (cache.Entries.TryGetValue(actionDescriptor, out cacheEntry))
             {
-                filters = GetFilters(controllerContext, cacheEntry.FilterItems);
+                // Deep copy the cached filter items as filter providers could modify them
+                var filterItems = new List<FilterItem>();
+                for (var i = 0; i < cacheEntry.FilterItems.Count; i++)
+                {
+                    var filterItem = cacheEntry.FilterItems[i];
+                    filterItems.Add(
+                        new FilterItem(filterItem.Descriptor)
+                        {
+                            Filter = filterItem.Filter,
+                            IsReusable = filterItem.IsReusable
+                        });
+                }
+
+                filters = GetFilters(controllerContext, filterItems);
```

Comment:
We should also remove the defensive copy inside `GetFilters` and inline that into the other call-site.